### PR TITLE
Fix issues with group functions

### LINKED ISSUE #2408: ASSERT: "!uuid.isNull()" in file Group.cpp, line 584
1. The database was locked
1. I've entered a wrong password and hit enter
1. I've entered the right password, used show password and hit enter
1. keepassxc crashed.

I still have the coredump of the crash.

## Backtrace

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
[Current thread is 1 (Thread 0x7f28deb7a640 (LWP 26665))]
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
        set = {__val = {0, 280375465148160, 549755814016, 549755814016, 1095216660735, 1095216660735, 0, 0, 0, 0, 0, 0, 88, 4393321634018672384, 140720354000688, 139813581952640}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007f28e2cb02f1 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0 <repeats 11 times>, 3, 140720354001088, 94350126039376, 139813585928408, 94350139144176}}, sa_flags = -742867200, sa_restorer = 0x7ffc02b67fd8}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007f28e31ce987 in qt_message_fatal (context=..., message=<synthetic pointer>...) at global/qlogging.cpp:1842
No locals.
#3  QMessageLogger::fatal (this=this@entry=0x7ffc02b680c0, msg=msg@entry=0x7f28e3470ec0 "ASSERT: \"%s\" in file %s, line %d") at global/qlogging.cpp:880
        message = <optimized out>
        ap = {{gp_offset = 40, fp_offset = 48, overflow_arg_area = 0x7ffc02b680c0, reg_save_area = 0x7ffc02b68000}}
#4  0x00007f28e31cde09 in qt_assert (assertion=assertion@entry=0x55cf99d3cf38 "!uuid.isNull()", file=file@entry=0x55cf99d3dde8 "/home/joerg/git/keepassxc/src/core/Group.cpp", line=line@entry=584) at ../../include/QtCore/../../src/corelib/global/qlogging.h:91
No locals.
#5  0x000055cf99bb331e in Group::findEntryByUuid (this=0x55cf9af50560, uuid=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qpointer.h:91
No locals.
#6  0x000055cf99c5cf25 in DatabaseWidget::restoreGroupEntryFocus (this=this@entry=0x55cf9a94af60, groupUuid=..., entryUuid=...) at /home/joerg/git/keepassxc/src/gui/DatabaseWidget.cpp:1348
        entry = <optimized out>
        group = 0x55cf9af50560
#7  0x000055cf99c5e459 in DatabaseWidget::unlockDatabase (this=0x55cf9a94af60, accepted=<optimized out>) at /home/joerg/git/keepassxc/src/gui/DatabaseWidget.cpp:881
        db = <optimized out>
#8  0x000055cf99bf4b91 in DatabaseWidget::qt_static_metacall (_o=0x55cf9a94af60, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at /home/joerg/git/keepassxc/build/src/keepassx_core_autogen/DMHXEJ42XS/moc_DatabaseWidget.cpp:469
        _t = 0x55cf9a94af60
#9  0x00007f28e33e125b in QMetaObject::activate(QObject*, int, int, void**) () at kernel/qobject.cpp:3771
        methodIndex = <optimized out>
        method_relative = <optimized out>
        callFunction = 0x55cf99bf4150 <DatabaseWidget::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>
        receiver = 0x55cf9a94af60
        receiverInSameThread = <optimized out>
        sw = {receiver = 0x55cf9a94af60, previousSender = 0x0, currentSender = {sender = 0x55cf9aeb3f30, signal = 7, ref = 1}, switched = true}
        c = 0x55cf9af31190
        last = 0x55cf9af31190
        locker = {val = 139813590424536}
        connectionLists = {connectionLists = 0x55cf9af311e0}
        list = <optimized out>
        currentThreadId = 0x7f28deb7a640
        signal_index = 7
        empty_argv = {0x0}
#10 0x00007f28e33e1877 in QMetaObject::activate (sender=sender@entry=0x55cf9aeb3f30, m=m@entry=0x55cf99d9e780 <DatabaseOpenWidget::staticMetaObject>, local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x7ffc02b682e0) at kernel/qobject.cpp:3633
No locals.
#11 0x000055cf99bebb7f in DatabaseOpenWidget::editFinished (this=this@entry=0x55cf9aeb3f30, _t1=<optimized out>, _t1@entry=true) at /home/joerg/git/keepassxc/build/src/keepassx_core_autogen/DMHXEJ42XS/moc_DatabaseOpenWidget.cpp:178
        _a = {0x0, 0x7ffc02b682dc}
#12 0x000055cf99c510d5 in DatabaseOpenWidget::openDatabase() () at /home/joerg/git/keepassxc/src/gui/DatabaseOpenWidget.cpp:225
        reader = {m_saveXml = false, m_error = false, m_errorStr = {static null = {<No data fields>}, d = 0x7f28e3473660 <QArrayData::shared_null>}, m_reader = {value = 0x55cf9aeb7b90, d = 0x55cf9a9ebab0}, m_version = 262144}
        masterKey = {value = 0x55cf9afcf010, d = 0x55cf9afcf000}
        file = {<QFileDevice> = {<QIODevice> = {<QObject> = {_vptr.QObject = 0x7f28e363b418 <vtable for QFile+16>, static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e35383c0 <qt_meta_stringdata_QObject>, data = 0x7f28e35382a0 <qt_meta_data_QObject>, static_metacall = 0x7f28e33e83d0 <QObject::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d_ptr = {d = 0x55cf9aeb8e80}, static staticQtMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e353b180 <qt_meta_stringdata_Qt>, data = 0x7f28e35384e0 <qt_meta_data_Qt>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e3636a40 <QObject::staticMetaObject>, stringdata = 0x7f28e3523aa0 <qt_meta_stringdata_QIODevice>, data = 0x7f28e35239a0 <qt_meta_data_QIODevice>, static_metacall = 0x7f28e33034e0 <QIODevice::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e363b880 <QIODevice::staticMetaObject>, stringdata = 0x7f28e3523900 <qt_meta_stringdata_QFileDevice>, data = 0x7f28e35238c0 <qt_meta_data_QFileDevice>, static_metacall = 0x7f28e32fcc80 <QFileDevice::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e363b6e0 <QFileDevice::staticMetaObject>, stringdata = 0x7f28e3523740 <qt_meta_stringdata_QFile>, data = 0x7f28e3523700 <qt_meta_data_QFile>, static_metacall = 0x7f28e32f9ec0 <QFile::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}
#13 0x000055cf99bf02d9 in DatabaseOpenWidget::qt_static_metacall (_o=<optimized out>, _id=<optimized out>, _a=<optimized out>, _c=<optimized out>) at /home/joerg/git/keepassxc/build/src/keepassx_core_autogen/DMHXEJ42XS/moc_DatabaseOpenWidget.cpp:115
        _t = <optimized out>
        result = <optimized out>
#14 0x00007f28e33e125b in QMetaObject::activate(QObject*, int, int, void**) () at kernel/qobject.cpp:3771
        methodIndex = <optimized out>
        method_relative = <optimized out>
        callFunction = 0x55cf99bf0350 <DatabaseOpenWidget::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>
        receiver = 0x55cf9aeb3f30
        receiverInSameThread = <optimized out>
        sw = {receiver = 0x55cf9aeb3f30, previousSender = 0x0, currentSender = {sender = 0x55cf9aec4e40, signal = 8, ref = 1}, switched = true}
        c = 0x55cf9aedf180
        last = 0x55cf9aedf180
        locker = {val = 139813590424760}
        connectionLists = {connectionLists = 0x55cf9aec7040}
        list = <optimized out>
        currentThreadId = 0x7f28deb7a640
        signal_index = 8
        empty_argv = {0x0}
#15 0x00007f28e33e1877 in QMetaObject::activate (sender=sender@entry=0x55cf9aec4e40, m=m@entry=0x7f28e4b5cce0 <QDialogButtonBox::staticMetaObject>, local_signal_index=local_signal_index@entry=1, argv=argv@entry=0x0) at kernel/qobject.cpp:3633
No locals.
#16 0x00007f28e480ad13 in QDialogButtonBox::accepted (this=this@entry=0x55cf9aec4e40) at .moc/moc_qdialogbuttonbox.cpp:281
No locals.
#17 0x00007f28e480b550 in QDialogButtonBoxPrivate::_q_handleButtonClicked (this=<optimized out>) at widgets/qdialogbuttonbox.cpp:864
        buttonRole = QDialogButtonBox::AcceptRole
        guard = <optimized out>
        button = 0x55cf9aec5300
        q = 0x55cf9aec4e40
#18 0x00007f28e33e125b in QMetaObject::activate(QObject*, int, int, void**) () at kernel/qobject.cpp:3771
        methodIndex = <optimized out>
        method_relative = <optimized out>
        callFunction = 0x7f28e480c5c0 <QDialogButtonBox::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>
        receiver = 0x55cf9aec4e40
        receiverInSameThread = <optimized out>
        sw = {receiver = 0x55cf9aec4e40, previousSender = 0x0, currentSender = {sender = 0x55cf9aec5300, signal = 9, ref = 1}, switched = true}
        c = 0x55cf9aeded80
        last = 0x55cf9aeded80
        locker = {val = 139813590424928}
        connectionLists = {connectionLists = 0x55cf9aec57c0}
        list = <optimized out>
        currentThreadId = 0x7f28deb7a640
        signal_index = 9
        empty_argv = {0x0}
#19 0x00007f28e33e1877 in QMetaObject::activate (sender=sender@entry=0x55cf9aec5300, m=m@entry=0x7f28e4b53bc0 <QAbstractButton::staticMetaObject>, local_signal_index=local_signal_index@entry=2, argv=argv@entry=0x7ffc02b68620) at kernel/qobject.cpp:3633
No locals.
#20 0x00007f28e476b702 in QAbstractButton::clicked (this=this@entry=0x55cf9aec5300, _t1=<optimized out>) at .moc/moc_qabstractbutton.cpp:308
        _a = {0x0, 0x7ffc02b6861c}
#21 0x00007f28e476b91a in QAbstractButtonPrivate::emitClicked (this=0x55cf9aec5340) at widgets/qabstractbutton.cpp:414
        q = 0x55cf9aec5300
        guard = <optimized out>
#22 0x00007f28e476c603 in QAbstractButton::click (this=0x55cf9aec5300) at widgets/qabstractbutton.cpp:871
        d = 0x55cf9aec5340
        guard = <optimized out>
        this = 0x55cf9aec5300
        d = <optimized out>
        guard = <optimized out>
        d = <optimized out>
        guard = <optimized out>
#23 0x000055cf99c68fbe in DialogyWidget::clickButton (this=this@entry=0x55cf9aeb3f30, standardButton=standardButton@entry=QDialogButtonBox::Ok) at /home/joerg/git/keepassxc/src/gui/DialogyWidget.cpp:78
        buttonBox = <optimized out>
        i = 0
        pb = <optimized out>
        buttonBoxes = {<QListSpecialMethods<QDialogButtonBox*>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, <No data fields>}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x55cf9ab74a60}, d = 0x55cf9ab74a60}}
#24 0x000055cf99c6900b in DialogyWidget::keyPressEvent (this=0x55cf9aeb3f30, e=0x7ffc02b68c20) at /home/joerg/git/keepassxc/src/gui/DialogyWidget.cpp:42
No locals.
#25 0x00007f28e46c3de7 in QWidget::event (this=0x55cf9aeb3f30, event=0x7ffc02b68c20) at kernel/qwidget.cpp:8940
        k = 0x7ffc02b68c20
        res = <optimized out>
        d = 0x55cf9aeb5f20
#26 0x00007f28e4685491 in QApplicationPrivate::notify_helper (this=this@entry=0x55cf9a441030, receiver=receiver@entry=0x55cf9aeb3f30, e=e@entry=0x7ffc02b68c20) at kernel/qapplication.cpp:3727
        consumed = <optimized out>
#27 0x00007f28e468da59 in QApplication::notify(QObject*, QEvent*) () at kernel/qapplication.cpp:3121
        w = 0x55cf9aeb3f30
        gw = 0x0
        isWidget = true
        key = 0x7ffc02b68c20
        def = true
        pr = {wp = {d = 0x55cf9af15ab0, value = 0x55cf9aec0d00}}
        isGraphicsWidget = false
        d = <optimized out>
        res = false
        me = <optimized out>
#28 0x00007f28e33b8019 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at ../../include/QtCore/5.11.2/QtCore/private/../../../../../src/corelib/thread/qthread_p.h:307
        selfRequired = true
        result = false
        cbdata = {0x55cf9aec0d00, 0x7ffc02b68c20, 0x7ffc02b6895f}
        d = <optimized out>
        threadData = 0x55cf9a4411b0
        scopeLevelCounter = {threadData = 0x55cf9a4411b0}
#29 0x00007f28e33b8064 in QCoreApplication::forwardEvent (receiver=<optimized out>, event=event@entry=0x7ffc02b68c20, originatingEvent=originatingEvent@entry=0x0) at kernel/qcoreapplication.cpp:1063
No locals.
#30 0x00007f28e46dd22b in QWidgetWindow::handleKeyEvent (this=this@entry=0x55cf9a4c1190, event=event@entry=0x7ffc02b68c20) at kernel/qwidgetwindow.cpp:698
        receiver = <optimized out>
#31 0x00007f28e46e0e79 in QWidgetWindow::event (this=0x55cf9a4c1190, event=0x7ffc02b68c20) at kernel/qwidgetwindow.cpp:273
        event = 0x7ffc02b68c20
        this = 0x55cf9a4c1190
        state = <optimized out>
        ev = <optimized out>
        widgetEvent = <optimized out>
#32 0x00007f28e4685491 in QApplicationPrivate::notify_helper (this=this@entry=0x55cf9a441030, receiver=receiver@entry=0x55cf9a4c1190, e=e@entry=0x7ffc02b68c20) at kernel/qapplication.cpp:3727
        consumed = <optimized out>
#33 0x00007f28e468cad0 in QApplication::notify(QObject*, QEvent*) () at kernel/qapplication.cpp:3486
        w = <optimized out>
        extra = <optimized out>
        isProxyWidget = <optimized out>
        d = <optimized out>
        res = false
        me = <optimized out>
#34 0x00007f28e33b8019 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at ../../include/QtCore/5.11.2/QtCore/private/../../../../../src/corelib/thread/qthread_p.h:307
        selfRequired = true
        result = false
        cbdata = {0x55cf9a4c1190, 0x7ffc02b68c20, 0x7ffc02b68bcf}
        d = <optimized out>
        threadData = 0x55cf9a4411b0
        scopeLevelCounter = {threadData = 0x55cf9a4411b0}
#35 0x00007f28e40b46e8 in QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
No symbol table info available.
#36 0x00007f28e40b9d65 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
No symbol table info available.
#37 0x00007f28e40940eb in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
No symbol table info available.
#38 0x00007f28de93190b in ?? () from /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
No symbol table info available.
#39 0x00007f28e33b6ceb in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at ../../include/QtCore/../../src/corelib/global/qflags.h:140
        d = 0x55cf9af3a1a0
        locker = {val = 94350134743744}
        ref = {d = 0x55cf9af3a1a0, locker = @0x7ffc02b68d78, exceptionCaught = true}
        app = <optimized out>
#40 0x00007f28e33bee62 in QCoreApplication::exec() () at ../../include/QtCore/../../src/corelib/global/qflags.h:120
        threadData = 0x55cf9a4411b0
        eventLoop = {<QObject> = {_vptr.QObject = 0x7f28e363e3c8 <vtable for QEventLoop+16>, static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e35383c0 <qt_meta_stringdata_QObject>, data = 0x7f28e35382a0 <qt_meta_data_QObject>, static_metacall = 0x7f28e33e83d0 <QObject::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d_ptr = {d = 0x55cf9af3a1a0}, static staticQtMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e353b180 <qt_meta_stringdata_Qt>, data = 0x7f28e35384e0 <qt_meta_data_Qt>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e3636a40 <QObject::staticMetaObject>, stringdata = 0x7f28e3532e60 <qt_meta_stringdata_QEventLoop>, data = 0x7f28e3532e00 <qt_meta_data_QEventLoop>, static_metacall = 0x7f28e33b6a10 <QEventLoop::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}
        returnCode = <optimized out>
#41 0x000055cf99bb892a in main (argc=<optimized out>, argv=<optimized out>) at /home/joerg/git/keepassxc/src/main.cpp:139
        app = {<QApplication> = {<QGuiApplication> = {<No data fields>}, static staticMetaObject = {d = {superdata = 0x7f28e4510b40 <QGuiApplication::staticMetaObject>, stringdata = 0x7f28e49f9b00 <qt_meta_stringdata_QApplication>, data = 0x7f28e49f9980 <qt_meta_data_QApplication>, static_metacall = 0x7f28e468a220 <QApplication::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e4b4ea80 <QApplication::staticMetaObject>, stringdata = 0x55cf99d38a00 <qt_meta_stringdata_Application>, data = 0x55cf99d38900 <qt_meta_data_Application>, static_metacall = 0x55cf99beffd0 <Application::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, m_mainWindow = 0x7ffc02b68f60, m_unixSignalNotifier = 0x55cf9a484620, static unixSignalSocket = {6, 7}, m_alreadyRunning = false, m_lockFile = 0x55cf9a484f20, m_lockServer = <incomplete type>, m_socketName = {static null = {<No data fields>}, d = 0x7f28d8003a00}}
        parser = {d = 0x55cf9a485100}
        configOption = {d = {d = 0x55cf9a487510}}
        keyfileOption = {d = {d = 0x55cf9a479030}}
        pwstdinOption = {d = {d = 0x55cf9a4869f0}}
        parentWindowOption = {d = {d = 0x55cf9a488410}}
        helpOption = {d = {d = 0x55cf9a487f40}}
        versionOption = {d = {d = 0x55cf9a4884e0}}
        fileNames = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, <No data fields>}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7f28e34756c0 <QListData::shared_null>}, d = 0x7f28e34756c0 <QListData::shared_null>}}, <No data fields>}
        __PRETTY_FUNCTION__ = "int main(int, char**)"
        mainWindow = {<QMainWindow> = {<QWidget> = {<QObject> = {_vptr.QObject = 0x55cf99da3b58 <vtable for MainWindow+16>, static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e35383c0 <qt_meta_stringdata_QObject>, data = 0x7f28e35382a0 <qt_meta_data_QObject>, static_metacall = 0x7f28e33e83d0 <QObject::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, d_ptr = {d = 0x55cf9a47fa00}, static staticQtMetaObject = {d = {superdata = 0x0, stringdata = 0x7f28e353b180 <qt_meta_stringdata_Qt>, data = 0x7f28e35384e0 <qt_meta_data_Qt>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}}, <QPaintDevice> = {<No data fields>}, static staticMetaObject = {d = {superdata = 0x7f28e3636a40 <QObject::staticMetaObject>, stringdata = 0x7f28e49fd480 <qt_meta_stringdata_QWidget>, data = 0x7f28e49fcda0 <qt_meta_data_QWidget>, static_metacall = 0x7f28e46c2520 <QWidget::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, data = 0x55cf9a47fb30}, static staticMetaObject = {d = {superdata = 0x7f28e4b50080 <QWidget::staticMetaObject>, stringdata = 0x7f28e4a1ce60 <qt_meta_stringdata_QMainWindow>, data = 0x7f28e4a1cc80 <qt_meta_data_QMainWindow>, static_metacall = 0x7f28e47c9580 <QMainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x7f28e4b59970 <qt_meta_extradata_QMainWindow>, extradata = 0x0}}}, static staticMetaObject = {d = {superdata = 0x7f28e4b59940 <QMainWindow::staticMetaObject>, stringdata = 0x55cf99d33540 <qt_meta_stringdata_MainWindow>, data = 0x55cf99d32f00 <qt_meta_data_MainWindow>, static_metacall = 0x55cf99bf3190 <MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, static BaseWindowTitle = {static null = {<No data fields>}, d = 0x55cf9a43fb40}, m_ui = {d = 0x55cf9a48ec80}, m_actionMultiplexer = {m_currentObject = {wp = {d = 0x55cf9af305f0, value = 0x55cf9a94af60}}, m_connections = {<QListSpecialMethods<SignalMultiplexer::Connection>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, <No data fields>}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x55cf9a8f8010}, d = 0x55cf9a8f8010}}}, m_clearHistoryAction = 0x55cf9a8a4260, m_searchWidgetAction = 0x55cf9a59a1d0, m_lastDatabasesActions = 0x55cf9a8ac580, m_copyAdditionalAttributeActions = 0x55cf9a8a41c0, m_openDatabases = {<QList<QString>> = {<QListSpecialMethods<QString>> = {<No data fields>}, {p = {static shared_null = {ref = {atomic = {_q_value = {<std::__atomic_base<int>> = {static _S_alignment = 4, _M_i = -1}, <No data fields>}}}, alloc = 0, begin = 0, end = 0, array = {0x0}}, d = 0x7f28e34756c0 <QListData::shared_null>}, d = 0x7f28e34756c0 <QListData::shared_null>}}, <No data fields>}, m_inactivityTimer = 0x7f28d8007e00, m_touchIDinactivityTimer = 0x1c, m_countDefaultAttributes = 4, m_trayIcon = 0x7f28d8007e30, m_screenLockListener = 0x55cf9a8fd980, m_appExitCalled = false, m_appExiting = false}
        pwstdin = <optimized out>
        exitCode = <optimized out>
```

## Debug Info
KeePassXC - Version 2.4.0-snapshot
Build Type: Snapshot
Revision: b45f0e3e

Bibliotheken:
- Qt 5.11.2
- libgcrypt 1.8.3

Betriebssystem: Debian GNU/Linux buster/sid
CPU-Architektur: x86_64
Kernel: linux 4.19.0-rc3-amd64

Aktivierte Erweiterungen:
- Auto-Type
- Browser Integration
- SSH Agent
- YubiKey

### PR DESCRIPTION
**Pull request by droidmonkey**
## Description

Resolves #2408

Just some cleanup and streamlining in the Group class. Eliminated function that had undetermined results based on heuristics of the input (Group::findEntry). Aligned Group::findEntryByPath with Group::findGroupByPath.

Eliminated unnecessary Q_ASSERTS and integrated the checks into the code function instead.

**Recommendation:** rename Group::children() to Group::groups(). Children overshadows QObject function AND it is the only place where the name children is used, otherwise we just call them groups...

**Recommendation 2:** Look for other places where QString::isNull() is used and evaluate whether it makes sense. This function is for legacy purposes only and QString::isEmpty() is preferred by the Qt Docs.

## Types of changes

- ✅ Refactor of code

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/core/Group.cpp [hunk 1]
```diff
@@ -552,59 +548,47 @@ QList<Entry*> Group::entriesRecursive(bool includeHistoryItems) const
...
-
-    entry = findEntryByPath(entryId);
-    if (entry) {
-        return entry;
-    }
-
-    for (Entry* entry : entriesRecursive(false)) {
-        if (entry->title() == entryId) {
-            return entry;
+    if (!uuid.isNull()) {
```
Comment (phoerious) [inline src/core/Group.cpp]:
```
if (uuid.isNull()) {
    return nullptr;
}
```

File: src/core/Group.cpp [hunk 2]
```diff
@@ -552,59 +548,47 @@ QList<Entry*> Group::entriesRecursive(bool includeHistoryItems) const
...
+        if (entryPath == (basePath + entry->title())
+            || (!entryPath.startsWith("/") && entry->title() == entryPath)) {
             return entry;
         }
     }
 
-    for (Group* group : asConst(m_children)) {
-        Entry* entry = group->findEntryByPath(entryPath, basePath + group->name() + QString("/"));
+    for (Group* group : children()) {
+        Entry* entry = group->findEntryByPathRecursion(entryPath, basePath + group->name() + "/");
```
Comment (phoerious) [inline src/core/Group.cpp]: Recursive

File: src/core/Group.cpp [hunk 3]
```diff
@@ -728,10 +710,11 @@ QSet<QUuid> Group::customIconsRecursive() const
 
 Group* Group::findGroupByUuid(const QUuid& uuid)
 {
-    Q_ASSERT(!uuid.isNull());
-    for (Group* group : groupsRecursive(true)) {
-        if (group->uuid() == uuid) {
-            return group;
+    if (!uuid.isNull()) {
```
Comment (phoerious) [inline src/core/Group.cpp]: See above.

File: tests/TestGroup.cpp [hunk 4]
```diff
@@ -493,56 +493,63 @@ void TestGroup::testFindEntry()
...
 
     // A valid UUID that does not exist in this database.
-    entry = db->rootGroup()->findEntry(QString("febfb01ebcdf9dbd90a3f1579dc75281"));
+    entry = db->rootGroup()->findEntryByUuid(QUuid("febfb01ebcdf9dbd90a3f1579dc75281"));
     QVERIFY(entry == nullptr);
 
     // An invalid UUID.
-    entry = db->rootGroup()->findEntry(QString("febfb01ebcdf9dbd90a3f1579dc"));
+    entry = db->rootGroup()->findEntryByUuid(QUuid("febfb01ebcdf9dbd90a3f1579dc"));
+    QVERIFY(entry == nullptr);
```
Comment (phoerious) [inline tests/TestGroup.cpp]: `QVERIFY(!entry)`

File: tests/TestGroup.cpp [hunk 5]
```diff
@@ -493,56 +493,63 @@ void TestGroup::testFindEntry()
...
+
+    // Empty strings
+    entry = db->rootGroup()->findEntryByUuid({});
+    QVERIFY(entry == nullptr);
+
+    entry = db->rootGroup()->findEntryByPath({});
     QVERIFY(entry == nullptr);
```
Comment (phoerious) [inline tests/TestGroup.cpp]: Ditto.

Comment (phoerious) [inline tests/TestGroup.cpp]: Ditto.

File: src/core/Group.cpp (hunk 1 above)
Comment (droidmonkey) [reply src/core/Group.cpp]: Why? then this function would have three return statements (short circuit, in-loop, and fall-through). I don't agree with that approach.

File: src/core/Group.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/core/Group.cpp]: same comment, changing this will create three return statements unnecessarily

**Comment by droidmonkey** [comment]
changes made to short circuit return nullptr.

**Pull request merged**